### PR TITLE
AVRO-2645: Validate names in schemas in Ruby client

### PR DESCRIPTION
**Pull request by dorner**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-2645
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
* Test that invalid names fail schema parsing

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/ruby/lib/avro/schema.rb [hunk 1]
```diff
@@ -29,6 +29,8 @@ class Schema
     NAMED_TYPES_SYM     = Set.new(NAMED_TYPES.map(&:to_sym))
     VALID_TYPES_SYM     = Set.new(VALID_TYPES.map(&:to_sym))
 
+    NAME_REGEX = /^[A-Za-z_][A-Za-z0-9_.]*$/
```
Comment (RyanSkraba) [inline lang/ruby/lib/avro/schema.rb]: Hello!  This lets through names like `valid.9badPackage..9badName`.

For reference, the python regex is here: https://github.com/apache/avro/blob/f8afef1f98b1f27deb9e73ef804ed74ba2d20bbb/lang/py/src/avro/schema.py#L183

And Java scans names without a regex: https://github.com/RyanSkraba/avro/blob/1c5d3ff4ff78551edd4bab9824642d6918d3565d/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L1518

(It looks like both Java and python also let `9badPackage` part through, but not the `9badName`...)

Comment (dorner) [reply lang/ruby/lib/avro/schema.rb]: Main issue is that the Ruby Avro library seems to assume that the namespace can be part of the name and it teases them apart later on, which is why it allows periods. Not quite sure why. I can add a check that there aren't two periods in a row.

**Comment by RyanSkraba** [comment]
Thanks for the change!  This regex passed on the mailing list recently and looks pretty rigorous!  https://github.com/heetch/cue-schema/blob/bb31c583dd09b87a4f1a206290bd57be39b6382f/avro/schema/schema.cue#L6

What do you think?  If it looks good to you, I'll start checking the other implementations -- I'm not convinced that they're all on the same page for fully qualified names like `name="valid.9badPackage.GoodName"`

(Disclaimer: I'm OK with regex, and the Avro spec to a degree, but this is my first exposure to ruby!)

**Comment by dorner** [comment]
Hey @RyanSkraba! That one looks great - I was trying to figure something else out with lookaheads and lookbehinds and was completely overthinking the problem. :p 
https://image.slidesharecdn.com/regexpsecrets-130228042149-phpapp02/95/regexp-secrets-6-638.jpg?cb=1362026041
Updated the PR, let me know what you think!

**Comment by tjwp** [comment]
I'm on board with enforcing the spec for names by default, but this will be a breaking change for anybody that is currently using non-conforming names.

I think we should include an option for people to opt out of the name validation. This is similar to what we did for enforcing the validation of field defaults: https://github.com/apache/avro/pull/497/files#diff-895d8c5301b553629a5f6d95d9bcc0c7R36-R43

**Comment by dorner** [comment]
Good idea - I'll add this in.

**Comment by dorner** [comment]
@RyanSkraba @tjwp added the option - I basically just copied word for word what we did with default validation 😄

**Comment by tjwp** [comment]
Thanks! Looks good to me. I'm not seeing an option to review, since this was already approved -- but approved!

**Comment by dorner** [comment]
@RyanSkraba @tjwp any idea when this will be merged?

**Comment by Fokko** [comment]
@dorner Thanks for the ping. As long as the CI is red, there is no chance to merge this. Could you rebase onto the master branch as well?

**Comment by dorner** [comment]
huh, it was green last I checked. 😮 sure, I'll rebase.

**Comment by dorner** [comment]
@Fokko all green now!

**Comment by dorner** [comment]
Woot woot! 🎉

**Pull request merged**